scripts/question_answering/model.py

```diff
@@ -1,636 +1,727 @@
-        context_mask : NDArray
-            input tensor with shape `(batch_size, context_sequence_length)`
-        query_mask : NDArray
-            input tensor with shape `(batch_size, query_sequence_length)`
-        context_max_len : int
-        query_max_len : int
-
-        Returns
-        --------
-        return : NDArray
-            output tensor with shape `(batch_size, context_sequence_length, 4*hidden_size)`
-        """
-        context_mask = nd.expand_dims(context_mask, axis=-1)
-        query_mask = nd.expand_dims(query_mask, axis=1)
-
-        similarity = self._calculate_trilinear_similarity(
-            context, query, context_max_len, query_max_len, w4mlu, bias)
-
-        similarity_dash = nd.softmax(mask_logits(similarity, query_mask))
-        similarity_dash_trans = nd.transpose(nd.softmax(
-            mask_logits(similarity, context_mask), axis=1), axes=(0, 2, 1))
-        c2q = nd.batch_dot(similarity_dash, query)
-        q2c = nd.batch_dot(nd.batch_dot(
-            similarity_dash, similarity_dash_trans), context)
-        return nd.concat(context, c2q, context * c2q, context * q2c, dim=-1)
-
-    def _calculate_trilinear_similarity(self, context, query, context_max_len, query_max_len, w4mlu, bias):
-        """Implement the computation of trilinear similarity function.
-
-            refer https://github.com/NLPLearn/QANet/blob/master/layers.py#L505
-
-            The similarity function is:
-                    f(w, q) = W[w, q, w * q]
-            where w and q represent the word in context and query respectively, and * operator means hadamard product.
-
-        Parameters
-        -----------
-        context : NDArray
-            input tensor with shape `(batch_size, context_sequence_length, hidden_size)`
-        query : NDArray
-            input tensor with shape `(batch_size, query_sequence_length, hidden_size)`
-        context_max_len : int
-        context_max_len : int
-
-        Returns
-        --------
-        similarity_mat : NDArray
-            output tensor with shape `(batch_size, context_sequence_length, query_sequence_length)`
-        """
-
-        subres0 = nd.tile(self.w4c(context), [1, 1, query_max_len])
-        subres1 = nd.tile(nd.transpose(
-            self.w4q(query), axes=(0, 2, 1)), [1, context_max_len, 1])
-        subres2 = nd.batch_dot(w4mlu * context,
-                               nd.transpose(query, axes=(0, 2, 1)))
-        similarity_mat = subres0 + subres1 + subres2 + bias
-        return similarity_mat
+r"""
+This file contains QANet model and all used layers.
+"""
+import math
+
+import mxnet as mx
+from mxnet import gluon, nd
+from mxnet.initializer import MSRAPrelu, Normal, Uniform, Xavier
+from gluonnlp.model import (DotProductAttentionCell, Highway,
+                            MultiHeadAttentionCell)
+
+
+try:
+    from config import (LAYERS_DROPOUT, EMB_ENCODER_CONV_CHANNELS, NUM_HIGHWAY_LAYERS,

```

Comment:
I suggest we can use `from config import *` and use e.g. `config.LAYERS_DROPOUT` in the context. Similarly, we can use the same method for the other hyperparameters from the config in the other places.